kong/plugins/opentelemetry/handler.lua

```diff
@@ -0,0 +1,188 @@
+local new_tab = require "table.new"
+local http = require "resty.http"
+local clone = require "table.clone"
+local otlp = require "kong.plugins.opentelemetry.otlp"
+local propagation = require "kong.tracing.propagation"
+local tablepool = require "tablepool"
+
+local ngx = ngx
+local kong = kong
+local ngx_log = ngx.log
+local ngx_ERR = ngx.ERR
+local ngx_DEBUG = ngx.DEBUG
+local ngx_now = ngx.now
+local ngx_update_time = ngx.update_time
+local ngx_req = ngx.req
+local ngx_get_headers = ngx_req.get_headers
+local timer_at = ngx.timer.at
+local clear = table.clear
+local propagation_parse = propagation.parse
+local propagation_set = propagation.set
+local tablepool_release = tablepool.release
+local tablepool_fetch = tablepool.fetch
+local null = ngx.null
+local encode_traces = otlp.encode_traces
+local transform_span = otlp.transform_span
+
+local POOL_BATCH_SPANS = "KONG_OTLP_BATCH_SPANS"
+local _log_prefix = "[otel] "
+
+local OpenTelemetryHandler = {
+  VERSION = "0.1.0",
+  PRIORITY = 14,
+}
+
+local default_headers = {
+  ["Content-Type"] = "application/x-protobuf",
+}
+
+-- worker-level spans cache
+local spans_cache = new_tab(5000, 0)
+local headers_cache = setmetatable({}, { __mode = "k" })
+local last_run_cache = setmetatable({}, { __mode = "k" })
+
+local function http_export_request(conf, pb_data, headers)
+  local httpc = http.new()
+  httpc:set_timeouts(conf.connect_timeout, conf.send_timeout, conf.read_timeout)
+  local res, err = httpc:request_uri(conf.endpoint, {
+    method = "POST",
+    body = pb_data,
+    headers = headers,
+  })
+  if not res then
+    ngx_log(ngx_ERR, _log_prefix, "failed to send request: ", err)
+  end
+
+  if res and res.status ~= 200 then
+    ngx_log(ngx_ERR, _log_prefix, "response error: ", res.status, ", body: ", res.body)
+  end
+end
+
+local function http_export(premature, conf)
+  if premature then
+    return
+  end
+
+  local spans = spans_cache
+  local spans_n = #spans
+  if spans_n == 0 then
+    return
+  end
+
+  local start = ngx_now()
+
+  -- cache http headers
+  local headers = default_headers
+  if conf.headers then
+    headers = headers_cache[conf.headers]
+  end
+
+  if not headers then
+    headers = clone(default_headers)
+    if conf.headers and conf.headers ~= null then
+      for k, v in pairs(conf.headers) do
+        headers[k] = v and v[1]
+      end
+    end
+    headers_cache[conf.headers] = headers
+  end
+
+  -- batch send spans
+  local batch_spans = tablepool_fetch(POOL_BATCH_SPANS, conf.batch_span_count, 0)
+
+  for i = 1, spans_cache.n do
```

Comment:
It seems `spans_cache.n` is always not nil here. So why don't we use this instead of `local spans_n = #spans` at #l67?